### ISSUE TITLE
twolayerOpt can cause the browser tab to crash

### ISSUE THREAD

**Issue by josephsmith** [erikbrinkman/d3-dag#52]
Here is a minimal repro. I can reproduce this on a recent model of MacBook Pro in both Safari and Chrome:

```
<html lang="en">
<head>
  <meta charset="UTF-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
  <meta http-equiv="X-UA-Compatible" content="ie=edge"/>
  <title>Crash browser tab on Mac Chrome with D3 Dag</title>
   <script src="https://d3js.org/d3.v5.min.js"></script>
   <script src="https://unpkg.com/d3-dag@0.5.0"></script>
   <script>
      let sourceData2 = []
     // creating a lesser amount of nodes will no longer crash
      for (i = 0; i < 100; i++) {
        sourceData2.push({"id":`${i}`, parentIds:[]});
      }
      for (i = 0; i < 100; i++) {
        sourceData2.push({"id":`${i+100}`, parentIds:[`${i}`]});
      }

      layout = d3
        .sugiyama()
        .decross(
          d3
           .decrossTwoLayer()
           .order(
               // changing twolayerOpt to twolayerMedian will no longer crash
              //d3.twolayerMedian()
              d3.twolayerOpt()
           )
         )
      var result = layout(d3.dagStratify()(sourceData2))
      console.log(result)
   </script>
</head>
<body>
</body>
</html>
```

**Comment by erikbrinkman** [comment]
**tl;dr** don't use twolayerOpt

Thanks for the minimal repro! This is somewhat expected as you're asking it to solve a problem with 100 factorial unique solutions, 2^9900 solutions the way the problem is structured. This specific example is pretty easy to optimize, so potentially there are better ways to set up the MILP that use fewer variables or take advantage of the MILP to better solve these problems, or use a better solver for this class of problems. A quick, google search didn't turn anything up, but I'm sure I've missed something.

I'm interested in making opt faster, so if you can point to, or come up with a better way of computing it, I'd implement it. Unfortunately, all I could find is an ILP solver, and the way to form this as an ILP is inherently bad.

**Comment by josephsmith** [comment]
Understood. I’m very unfamiliar with this category of problem and
admittedly glossed over the algorithmic complexity details (I was using
Prefect and this came up).

I do appreciate your prompt response.

I suppose my only other question is whether it is possible to trivially
detect non viable graph topologies/sizes and throw an exception rather than
allowing the browser to crash. It’s entirely possible that my suggestion is
naive, and if so, I apologize in advance.

**Comment by erikbrinkman** [comment]
Not at all, that seems like a reasonable thing to have, especially for this operator. I'll throw up a pr and you can tell me if it looks reasonable.

**Comment by erikbrinkman** [comment]
Also, I looked into trying to reduce the number of variables for the MILP and I think the way it's currently written is the best you can do under LP constraints, but it takes the number of states up to `2^(width * width)` where as if you just model permutations it's only `2 ^ (width * log2(width))`, this makes me think that an optimizer more tailored to this problem, will perform much better.
